Sort by Recently Used

tomershvueli opened the pull request:
Hey @virejdasani ! 

This is a PR to support the following improvements that we discussed over email: 
- When filtering, the search query will also be run against the emoji names, not just keywords. This helps for things like flag names where filtering by country wouldn't find the flag you were looking for
- When a user hits enter, it will immediately select the top most emoji, even if not explicitly selected. I found this to be amazing in the user's performance of quickly getting to where they need to go. You can even start searching and hit enter when you see your preferred emoji at the top, without having to hit 'down'
- The vast majority of the code here is to support a 'recently used' feature. Right now the list contains up to 10 items, and those items will show up first in the modal window. One can of course still search and filter through, but by default, the most recently used will show up near the top. 
  - One thing to note is that the recently used I believe will be wiped out when the app restarts. If we would want to persist this across app runs, we would need to write the LRU Map object to disk or something similar probably. 
  - I also wasn't able to test this functionality (or the others come to think of it... 😅) on Windows or Linux, so would be best to test it out to make sure all is working well. 

Happy to hear your thoughts! 

No rush on this since I know you're busy.

tomershvueli committed:
feat: Also search through emoji names when filtering emoji suggestions

tomershvueli committed:
feat: When user hits enter, autoclick on first item in list

tomershvueli committed:
feat: Add an LRU Map to keep track of most recently used 10 items and sort those first

tomershvueli committed:
feat: Bring recents functionality to windowsNlinux

tomershvueli committed:
fix: Fix indentation

tomershvueli committed:
fix: No dev mode :)

tomershvueli commented on app/macos/main.js:
```diff
@@ -105,6 +112,36 @@ ipcMain.on("typeEmoji", (_event, arg) => {
 });
 
+// Return filtered and sorted emojis based on a search query
+ipcMain.handle("getEmojisForSearchString", (_event, arg) => {
```
Since the LRU Map is defined in main.js instead of index.js, it made sense to do the filtering and sorting here and pass along that array to index.js.

tomershvueli commented on app/macos/main.js:
```diff
@@ -16,9 +17,15 @@ const path = require("path");
 let window = undefined;
 
+// Let's preset our LRU Map that we'll use to keep track of recent uses
+let lruMap = new LRUMap(10);
```
We want the LRU Map in main.js instead of index.js to persist the object across its entire runtime, not restart every time the modal pops up.

virejdasani commented:
@tomershvueli, this looks absolutely amazing! It works flawlessly on my MacBook. I'll test it on Windows and Linux then go ahead with the merge.
Also, to persist data, we can just set the recent emoji data to [`localStorage`](https://www.saintsatplay.com/blog/2015-10-21-storing-and-retrieving-objects-with-localstorage) and then every time the app is restarted, we can use `localStorage.getItem`

virejdasani commented:
@tomershvueli, I tried doing `npm start` on my MacBook in the `windowsNlinux` folder and there are a lot of errors. Could you check it out please, thanks!

tomershvueli committed:
feat: Use store to save user's recents between app runs

tomershvueli committed:
fix: Fix windowsNlinux version, remove no longer necessary emojis.js import in html page

tomershvueli committed:
feat: Bring store to windowsNlinux version

tomershvueli commented:
@virejdasani take a look after the latest few pulls, I believe I fixed the `windowsNlinux` hierarchy and added a store to both as well to save recents in between app runs!

virejdasani commented:
This looks awesome and works perfectly across all operating systems.
Going ahead with the merge, thanks a lot!
Also, just wanted your opinion on whether I should replace the `Ctrl + E` shortcut with the `:` (colon key) to make Geniemoji work more like emoji in Discord, Slack, or even here in GitHub?

tomershvueli commented:
> Also, just wanted your opinion on whether I should replace the `Ctrl + E` shortcut with the `:` (colon key) to make Geniemoji work more like emoji in Discord, Slack, or even here in GitHub?

This is an interesting thought! I would be cautious because the colon key can also be used regularly without trying to type emojis sometimes. Would you be able to trigger it using a colon followed by any letter of the alphabet or a `+` key? That way if I type a colon followed by a space it wouldn't trigger the modal, but would in other cases.

virejdasani commented:
> This is an interesting thought! I would be cautious because the colon key can also be used regularly without trying to type emojis sometimes. Would you be able to trigger it using a colon followed by any letter of the alphabet or a `+` key? That way if I type a colon followed by a space it wouldn't trigger the modal, but would in other cases.

I'll try to make it so that if there is a letter typed after the colon, then it triggers Geniemoji and automatically writes that letter in the search box so `: [space]` doesn't trigger anything. What do you think?

tomershvueli commented:
> I'll try to make it so that if there is a letter typed after the colon, then it triggers Geniemoji and automatically writes that letter in the search box so `: [space]` doesn't trigger anything. What do you think?

I think that sounds awesome! Would love to see that in action!

virejdasani commented:
> I think that sounds awesome! Would love to see that in action!

@tomershvueli, today was a holiday so I had some free time to implement the feature. It's in the master branch right now! I haven't packaged the app yet so you'll have to `npm install` for now. Check it out and let me know what your thoughts are!

tomershvueli commented:
@virejdasani neat! 2 things that might make this just perfect: 
- If I type in `:s` can the `s` already be prepopulated in the modal window? 
- More importantly, if I type in `:s`, find my emoji and hit 'enter', can the emoji I selected *replace* the `:s` that I typed in? For example, right now it would output `:s🤩` where I just want the emoji. How plausible is this?

virejdasani commented:
Thanks, I'm working on it now, I'll let you know once it's done 👍

virejdasani commented:
@tomershvueli, It finally works as expected now! Check it out and let me know, thanks a lot!

tomershvueli commented:
Looks great! 
Some small notes: 
- Change keyName to a `+` instead of `=` for code `13` in keys.js, since the modal only triggers for a `+` and not for an `=`
- Let's manually call `search()` once in index.js on line 111 to trigger a search when the modal first pops up and gets populated with a letter, otherwise the user has to keep typing to try and filter for their emoji
- The `letterToWrite` code in index.js can be simplified using the [`Array.find` method](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find). Something like this: ```const key = keysArr.find(i => i.keyCode === letterCode); if (key) ...```

Besides that, this is looking great!

virejdasani commented:
Thanks a bunch!

virejdasani commented:
I decided to hold off on the colon feature because it was presenting too many bugs and entirely not working on windows and linux. I think ctlr+e is fine for now. Maybe in a future update, I'll implement this feature. Anyways, thanks a lot for your help!

tomershvueli commented:
@virejdasani I totally understand that, as this is starting to get into the territory of implementing things that Electron wasn't exactly built for... happy to give any input as needed in the future!

virejdasani commented:
Thanks a lot :)
